external/js/cothority/src/network/connection.ts

```diff
@@ -191,67 +188,66 @@ export class WebSocketConnection implements IConnection {
     }
 
     /** @inheritdoc */
-    sendStream<T extends Message>(message: Message, reply: typeof Message,
-                                  onMessage: (data: T, ws: WebSocketAdapter) => void,
-                                  onClose: (code: number, reason: string) => void,
-                                  onError: (err: Error) => void): WebSocketAdapter {
+    sendStream<T extends Message>(message: Message, reply: typeof Message):
+        Observable<[T, WebSocketAdapter]> {
+        return new Observable((sub) => {
 
-        if (!message.$type) {
-            onError(new Error(`message "${message.constructor.name}" is not registered`));
-            return;
-        }
+            if (!message.$type) {
+                sub.error(new Error(`message "${message.constructor.name}" is not registered`));
+                return;
+            }
 
-        if (!reply.$type) {
-            onError(new Error(`message "${reply}" is not registered`));
-            return;
-        }
+            const path = this.getURL() + "/" + this.service + "/" + message.$type.name.replace(/.*\./, "");
+            Log.lvl4(`Socket: new WebSocket(${path})`);
+            const ws = factory(path);
+            const bytes = Buffer.from(message.$type.encode(message).finish());
+            const timer = setTimeout(() => ws.close(1000, "timeout"), this.timeout);
+            ws.onOpen(() => {
+                Log.lvl3("Sending message to", path);
+                ws.send(bytes);
+            });
 
-        const path = `${this.getURL()}/${this.service}/${message.$type.name.replace(/.*\./, "")}`;
-        Log.lvl4(`Socket: new WebSocket(${path})`);
-        const ws = factory(path);
-        const bytes = Buffer.from(message.$type.encode(message).finish());
-        const timer = setTimeout(() => ws.close(1000, "timeout"), this.timeout);
-        ws.onOpen(() => {
-            Log.lvl3("Sending message to", path);
-            ws.send(bytes);
-        });
+            ws.onMessage((data: Buffer) => {
+                clearTimeout(timer);
+                const buf = Buffer.from(data);
```

Comment:
Hum, `data` is already of type `Buffer`, why `Buffer.from` it?

Reply:
I think I remember that one. I'm not sure that all the `nativescript` and other websocket implementations always return a Buffer here. I prefer keeping it here...